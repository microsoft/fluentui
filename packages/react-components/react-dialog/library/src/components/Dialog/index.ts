export { Dialog } from './Dialog';
export type {
  DialogContextValues,
  DialogModalType,
  DialogOpenChangeData,
  DialogOpenChangeEvent,
  DialogOpenChangeEventHandler,
  DialogProps,
  DialogSlots,
  DialogState,
  InternalDialogSlots,
} from './Dialog.types';
export { renderDialog_unstable } from './renderDialog';
export { useDialog_unstable } from './useDialog';
export { useDialogContextValues_unstable } from './useDialogContextValues';
