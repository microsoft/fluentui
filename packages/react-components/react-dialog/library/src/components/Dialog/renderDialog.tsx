/** @jsxRuntime automatic */
/** @jsxImportSource @fluentui/react-jsx-runtime */

import { assertSlots } from '@fluentui/react-utilities';
import * as React from 'react';

import { DialogProvider, DialogSurfaceProvider } from '../../contexts';
import type { DialogState, DialogContextValues, DialogSlots } from './Dialog.types';

/**
 * Render the final JSX of Dialog
 */
export const renderDialog_unstable = (state: DialogState, contextValues: DialogContextValues) => {
  assertSlots<DialogSlots>(state);

  return (
    <DialogProvider value={contextValues.dialog}>
      <DialogSurfaceProvider value={contextValues.dialogSurface}>
        {state.trigger}
        {state.content && (
          <state.surfaceMotion>
            {/* Casting here as content should be equivalent to <DialogSurface/> */}
            {/* FIXME: content should not be ReactNode it should be ReactElement instead. */}
            {state.content as React.ReactElement}
          </state.surfaceMotion>
        )}
      </DialogSurfaceProvider>
    </DialogProvider>
  );
};
