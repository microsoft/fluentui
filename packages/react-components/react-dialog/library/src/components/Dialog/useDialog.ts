import * as React from 'react';
import { useControllableState, useEventCallback, useId } from '@fluentui/react-utilities';
import { useHasParentContext } from '@fluentui/react-context-selector';
import { useFocusFirstElement } from '../../utils';
import { DialogContext } from '../../contexts';

import type { DialogOpenChangeData, DialogProps, DialogState } from './Dialog.types';
import { useModalAttributes } from '@fluentui/react-tabster';

/**
 * Create the state required to render Dialog.
 *
 * The returned state can be modified with hooks such as useDialogStyles_unstable,
 * before being passed to renderDialog_unstable.
 *
 * @param props - props from this instance of Dialog
 */
export const useDialog_unstable = (props: DialogProps): DialogState => {
  const { children, modalType = 'modal', onOpenChange, inertTrapFocus = false } = props;

  const [trigger, content] = childrenToTriggerAndContent(children);

  const [open, setOpen] = useControllableState({
    state: props.open,
    defaultState: props.defaultOpen,
    initialState: false,
  });

  const requestOpenChange = useEventCallback((data: DialogOpenChangeData) => {
    onOpenChange?.(data.event, data);

    // if user prevents default then do not change state value
    // otherwise updates state value and trigger reference to the element that caused the opening
    if (!data.event.isDefaultPrevented()) {
      setOpen(data.open);
    }
  });

  const focusRef = useFocusFirstElement(open, modalType);

  const { modalAttributes, triggerAttributes } = useModalAttributes({
    trapFocus: modalType !== 'non-modal',
    legacyTrapFocus: !inertTrapFocus,
  });

  const isNestedDialog = useHasParentContext(DialogContext);

  return {
    components: {
      backdrop: 'div',
    },
    inertTrapFocus,
    open,
    modalType,
    content,
    trigger,
    requestOpenChange,
    dialogTitleId: useId('dialog-title-'),
    isNestedDialog,
    dialogRef: focusRef,
    modalAttributes,
    triggerAttributes,
  };
};

/**
 * Extracts trigger and content from children
 */
function childrenToTriggerAndContent(
  children: React.ReactNode,
): readonly [trigger: React.ReactNode, content: React.ReactNode] {
  const childrenArray = React.Children.toArray(children) as React.ReactElement[];
  if (process.env.NODE_ENV !== 'production') {
    if (childrenArray.length !== 1 && childrenArray.length !== 2) {
      // eslint-disable-next-line no-console
      console.warn(/* #__DE-INDENT__ */ `
        @fluentui/react-dialog [useDialog]:
        Dialog must contain at least one child <DialogSurface/>,
        and at most two children <DialogTrigger/> <DialogSurface/> (in this order).
      `);
    }
  }
  switch (childrenArray.length) {
    // case where there's a trigger followed by content
    case 2:
      return childrenArray as [trigger: React.ReactNode, content: React.ReactNode];
    // case where there's only content
    case 1:
      return [undefined, childrenArray[0]];
    // unknown case
    default:
      return [undefined, undefined];
  }
}
