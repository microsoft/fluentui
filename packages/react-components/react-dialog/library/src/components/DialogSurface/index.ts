export { DialogSurface } from './DialogSurface';
export type {
  DialogSurfaceContextValues,
  DialogSurfaceElement,
  DialogSurfaceProps,
  DialogSurfaceSlots,
  DialogSurfaceState,
} from './DialogSurface.types';
export { renderDialogSurface_unstable } from './renderDialogSurface';
export { useDialogSurface_unstable } from './useDialogSurface';
export { dialogSurfaceClassNames, useDialogSurfaceStyles_unstable } from './useDialogSurfaceStyles.styles';
export { useDialogSurfaceContextValues_unstable } from './useDialogSurfaceContextValues';
