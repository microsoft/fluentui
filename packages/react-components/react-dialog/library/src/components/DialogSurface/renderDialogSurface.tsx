/** @jsxRuntime automatic */
/** @jsxImportSource @fluentui/react-jsx-runtime */

import { Portal } from '@fluentui/react-portal';
import { assertSlots } from '@fluentui/react-utilities';

import { DialogSurfaceProvider } from '../../contexts';
import type { DialogSurfaceState, DialogSurfaceSlots, DialogSurfaceContextValues } from './DialogSurface.types';

/**
 * Render the final JSX of DialogSurface
 */
export const renderDialogSurface_unstable = (state: DialogSurfaceState, contextValues: DialogSurfaceContextValues) => {
  assertSlots<DialogSurfaceSlots>(state);

  return (
    <Portal mountNode={state.mountNode}>
      {state.backdrop && (
        <state.backdropMotion>
          <state.backdrop />
        </state.backdropMotion>
      )}
      <DialogSurfaceProvider value={contextValues.dialogSurface}>
        <state.root />
      </DialogSurfaceProvider>
    </Portal>
  );
};
