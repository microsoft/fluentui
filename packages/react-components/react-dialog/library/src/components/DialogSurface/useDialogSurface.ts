import { Escape } from '@fluentui/keyboard-keys';
import { presenceMotionSlot, type PresenceMotionSlotProps } from '@fluentui/react-motion';
import {
  useEventCallback,
  useMergedRefs,
  isResolvedShorthand,
  slot,
  getIntrinsicElementProps,
  useIsomorphicLayoutEffect,
} from '@fluentui/react-utilities';
import * as React from 'react';

import { useDialogContext_unstable } from '../../contexts';
import { useDisableBodyScroll } from '../../utils/useDisableBodyScroll';
import { DialogBackdropMotion } from '../DialogBackdropMotion';
import type { DialogSurfaceElement, DialogSurfaceProps, DialogSurfaceState } from './DialogSurface.types';

/**
 * Create the state required to render DialogSurface.
 *
 * The returned state can be modified with hooks such as useDialogSurfaceStyles_unstable,
 * before being passed to renderDialogSurface_unstable.
 *
 * @param props - props from this instance of DialogSurface
 * @param ref - reference to root HTMLElement of DialogSurface
 */
export const useDialogSurface_unstable = (
  props: DialogSurfaceProps,
  ref: React.Ref<DialogSurfaceElement>,
): DialogSurfaceState => {
  const modalType = useDialogContext_unstable(ctx => ctx.modalType);
  const isNestedDialog = useDialogContext_unstable(ctx => ctx.isNestedDialog);

  const modalAttributes = useDialogContext_unstable(ctx => ctx.modalAttributes);
  const dialogRef = useDialogContext_unstable(ctx => ctx.dialogRef);
  const requestOpenChange = useDialogContext_unstable(ctx => ctx.requestOpenChange);
  const dialogTitleID = useDialogContext_unstable(ctx => ctx.dialogTitleId);
  const open = useDialogContext_unstable(ctx => ctx.open);

  const handledBackdropClick = useEventCallback((event: React.MouseEvent<HTMLDivElement>) => {
    if (isResolvedShorthand(props.backdrop)) {
      props.backdrop.onClick?.(event);
    }
    if (modalType === 'modal' && !event.isDefaultPrevented()) {
      requestOpenChange({
        event,
        open: false,
        type: 'backdropClick',
      });
    }
  });

  const handleKeyDown = useEventCallback((event: React.KeyboardEvent<HTMLDivElement>) => {
    props.onKeyDown?.(event);

    if (event.key === Escape && !event.isDefaultPrevented()) {
      requestOpenChange({
        event,
        open: false,
        type: 'escapeKeyDown',
      });
      // stop propagation to avoid conflicting with other elements that listen for `Escape`
      // e,g: nested Dialog, Popover, Menu and Tooltip
      event.preventDefault();
    }
  });

  const backdrop = slot.optional(props.backdrop, {
    renderByDefault: modalType !== 'non-modal',
    defaultProps: {
      'aria-hidden': 'true',
    },
    elementType: 'div',
  });
  if (backdrop) {
    backdrop.onClick = handledBackdropClick;
  }

  const { disableBodyScroll, enableBodyScroll } = useDisableBodyScroll();

  useIsomorphicLayoutEffect(() => {
    if (isNestedDialog || modalType === 'non-modal') {
      return;
    }

    disableBodyScroll();

    return () => {
      enableBodyScroll();
    };
  }, [enableBodyScroll, isNestedDialog, disableBodyScroll, modalType]);

  return {
    components: {
      backdrop: 'div',
      root: 'div',
      // TODO: remove once React v18 slot API is modified
      // This is a problem at the moment due to UnknownSlotProps assumption
      // that `children` property is `ReactNode`, which in this case is not valid
      // as PresenceComponentProps['children'] is `ReactElement`
      backdropMotion: DialogBackdropMotion as React.FC<PresenceMotionSlotProps>,
    },
    open,
    backdrop,
    isNestedDialog,
    mountNode: props.mountNode,
    root: slot.always(
      getIntrinsicElementProps('div', {
        tabIndex: -1, // https://github.com/microsoft/fluentui/issues/25150
        'aria-modal': modalType !== 'non-modal',
        role: modalType === 'alert' ? 'alertdialog' : 'dialog',
        'aria-labelledby': props['aria-label'] ? undefined : dialogTitleID,
        ...props,
        ...modalAttributes,
        onKeyDown: handleKeyDown,
        // FIXME:
        // `DialogSurfaceElement` is wrongly assigned to be `HTMLElement` instead of `HTMLDivElement`
        // but since it would be a breaking change to fix it, we are casting ref to it's proper type
        ref: useMergedRefs(ref, dialogRef) as React.Ref<HTMLDivElement>,
      }),
      { elementType: 'div' },
    ),
    backdropMotion: presenceMotionSlot(props.backdropMotion, {
      elementType: DialogBackdropMotion,
      defaultProps: {
        appear: true,
        visible: open,
      },
    }),

    // Deprecated properties
    transitionStatus: undefined,
  };
};
