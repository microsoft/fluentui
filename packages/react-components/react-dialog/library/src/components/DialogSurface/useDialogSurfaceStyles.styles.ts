import { makeResetStyles, makeStyles, mergeClasses } from '@griffel/react';
import type { SlotClassNames } from '@fluentui/react-utilities';
import { tokens } from '@fluentui/react-theme';
import { createFocusOutlineStyle } from '@fluentui/react-tabster';
import {
  FULLSCREEN_DIALOG_SCROLLBAR_OFFSET,
  MEDIA_QUERY_BREAKPOINT_SELECTOR,
  MEDIA_QUERY_SHORT_SCREEN,
  SURFACE_BORDER_WIDTH,
  SURFACE_PADDING,
} from '../../contexts';
import type { DialogSurfaceSlots, DialogSurfaceState } from './DialogSurface.types';

export const dialogSurfaceClassNames: SlotClassNames<DialogSurfaceSlots> = {
  root: 'fui-DialogSurface',
  backdrop: 'fui-DialogSurface__backdrop',
};

/**
 * Styles for the root slot
 */
const useRootBaseStyle = makeResetStyles({
  ...createFocusOutlineStyle(),
  inset: 0,
  padding: SURFACE_PADDING,
  margin: 'auto',
  borderStyle: 'none',
  overflow: 'unset',
  border: `${SURFACE_BORDER_WIDTH} solid ${tokens.colorTransparentStroke}`,
  borderRadius: tokens.borderRadiusXLarge,

  display: 'block',
  userSelect: 'unset',
  visibility: 'unset',
  position: 'fixed',
  height: 'fit-content',
  maxWidth: '600px',
  maxHeight: '100vh',
  boxSizing: 'border-box',
  backgroundColor: tokens.colorNeutralBackground1,
  color: tokens.colorNeutralForeground1,

  [MEDIA_QUERY_BREAKPOINT_SELECTOR]: {
    maxWidth: '100vw',
  },

  [MEDIA_QUERY_SHORT_SCREEN]: {
    overflowY: 'auto',
    // We need to offset the scrollbar by adding transparent borders otherwise
    // it conflicts with the border radius.
    paddingRight: `calc(${SURFACE_PADDING} - ${FULLSCREEN_DIALOG_SCROLLBAR_OFFSET})`,
    borderRightWidth: FULLSCREEN_DIALOG_SCROLLBAR_OFFSET,
    borderTopWidth: FULLSCREEN_DIALOG_SCROLLBAR_OFFSET,
    borderBottomWidth: FULLSCREEN_DIALOG_SCROLLBAR_OFFSET,
  },
});

const useBackdropBaseStyle = makeResetStyles({
  inset: '0px',
  backgroundColor: 'rgba(0, 0, 0, 0.4)',
  position: 'fixed',
});

const useBackdropStyles = makeStyles({
  nestedDialogBackdrop: {
    backgroundColor: tokens.colorTransparentBackground,
  },
});

/**
 * Apply styling to the DialogSurface slots based on the state
 */
export const useDialogSurfaceStyles_unstable = (state: DialogSurfaceState): DialogSurfaceState => {
  const { isNestedDialog, root, backdrop } = state;

  const rootBaseStyle = useRootBaseStyle();

  const backdropBaseStyle = useBackdropBaseStyle();
  const backdropStyles = useBackdropStyles();

  root.className = mergeClasses(dialogSurfaceClassNames.root, rootBaseStyle, root.className);

  if (backdrop) {
    backdrop.className = mergeClasses(
      dialogSurfaceClassNames.backdrop,
      backdropBaseStyle,
      isNestedDialog && backdropStyles.nestedDialogBackdrop,
      backdrop.className,
    );
  }

  return state;
};
