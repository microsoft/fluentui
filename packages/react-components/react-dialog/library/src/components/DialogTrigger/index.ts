export { DialogTrigger } from './DialogTrigger';
export type {
  DialogTriggerAction,
  DialogTriggerChildProps,
  DialogTriggerProps,
  DialogTriggerState,
} from './DialogTrigger.types';
export { renderDialogTrigger_unstable } from './renderDialogTrigger';
export { useDialogTrigger_unstable } from './useDialogTrigger';
