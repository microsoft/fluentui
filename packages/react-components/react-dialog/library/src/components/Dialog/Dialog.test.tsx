import * as React from 'react';
import { render } from '@testing-library/react';
import { Dialog } from './Dialog';
import { DialogProps } from './Dialog.types';
import { isConformant } from '../../testing/isConformant';
import { DialogTrigger } from '../DialogTrigger/DialogTrigger';
import { makeStyles, mergeClasses } from '@griffel/react';
import { DialogSurface, DialogSurfaceProps } from '../../DialogSurface';

describe('Dialog', () => {
  isConformant<DialogProps>({
    Component: Dialog,
    displayName: 'Dialog',
    disabledTests: [
      // Dialog does not render DOM elements
      'component-handles-ref',
      'component-has-root-ref',
      'component-handles-classname',
      'component-has-static-classnames-object',
      // TODO:
      // onOpenChange: A second (data) argument cannot be a union
      'consistent-callback-args',
      // Dialog does not have own styles
      'make-styles-overrides-win',
    ],
  });

  it('renders a default state', () => {
    const result = render(
      <Dialog>
        <div>Default Dialog</div>
      </Dialog>,
    );
    expect(result.container).toMatchSnapshot();
  });

  it('Testing DialogSurface with toBeVisible works as expected', () => {
    // eslint-disable-next-line @griffel/styles-file
    const useStyles = makeStyles({
      root: {
        left: '2px',
      },
    });
    const CustomDialogSurface = React.forwardRef<HTMLDivElement, DialogSurfaceProps>((props, ref) => {
      const styles = useStyles();
      return <DialogSurface ref={ref} className={mergeClasses(styles.root, props.className)} {...props} />;
    });

    const result = render(
      <Dialog>
        <DialogTrigger disableButtonEnhancement>
          <button data-testid="trigger">Open dialog</button>
        </DialogTrigger>
        <CustomDialogSurface>
          <div data-testid="surface-content">content in surface</div>
        </CustomDialogSurface>
      </Dialog>,
    );

    result.getByTestId('trigger').click();

    expect(result.getByTestId('surface-content')).toBeInTheDocument();
    expect(result.getByTestId('surface-content')).toBeVisible();
  });
});
