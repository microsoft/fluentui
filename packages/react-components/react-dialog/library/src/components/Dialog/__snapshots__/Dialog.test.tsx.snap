// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dialog renders a default state 1`] = `<div />`;
